src/pkcs11/slot.c

```diff
@@ -235,10 +236,10 @@ CK_RV card_detect(sc_reader_t *reader)
 		sc_log(context, "%s: Card changed", reader->name);
 		/* The following should never happen - but if it
 		 * does we'll be stuck in an endless loop.
-		 * So better be fussy.
-		if (!retry--)
-			return CKR_TOKEN_NOT_PRESENT; */
+		 * So better be fussy.*/
 		card_removed(reader);
+		if (!retry--)
+			return CKR_TOKEN_NOT_PRESENT;
```

Comment:
> The changes regarding slot handling remove the infinite loop from card_detect() when SC_READER_CARD_CHANGED is repeatedly acquired. It happened when the reader was removed & inserted back, so the timeout was reached, but no reconnection between the refresh_attributes() in pcsc-reader.c happens - the card handle is invalid, so the flag about changed is repeatedly set. This should probably be handled differently - any insight would be appreciated.

This was the initial problem, however I suppose it is really a rare case.

Reply:
If it is a siginificant problem even in rare cases, I don't mind having a prevention in the code, but as written above, the "prevention" currently doesn't work because `retry` is currently reset every time you call the function.

Reply:
Yes, I understand that it should be `static` to prevent the behavior in between calls to that function, but the problem I've run into was cycling just inside one call, since it was stucked here, calling `sc_detect_card_presence`, getting `SC_READER_CARD_CHANGED` and going back to `again`:

https://github.com/OpenSC/OpenSC/blob/f7070f4c9f9c6e1cb8d2efe838f73d54c78553da/src/pkcs11/slot.c#L221-L242

However, the only case I can hit this problem is now is with the oseid, where I'm trying to find out what also causes the problem described here https://github.com/OpenSC/OpenSC/pull/2803#issuecomment-2404556863. So I'll fix the `static` or remove the whole thing, if I'll find out what is causing it.